Prepare 0.23.0

ctz opened the pull request:
# Release notes (proposed)

- **Default cryptography provider changed** to [`aws-lc-rs`]. Note that this has some implications on [platform support and build-time tool requirements][aws-lc-rs-reqs] such as `cmake` on all platforms and `nasm` on Windows.
  Support for `ring` continues to be available: set the `ring` crate feature.

- **Support for FIPS validated mode** with [`aws-lc-rs`]: see [the manual section][fips-manual] and [aws-lc-rs's FIPS documentation][aws-fips-docs]. Note that `aws-lc-rs` in FIPS mode has further build-time requirements as detailed in the FIPS documentation.

- **Support for process-wide selection of `CryptoProvider`s**.  See [the documentation][process-provider]. Note that callers of `ClientConfig::builder()`, `ServerConfig::builder()`, `WebPkiServerVerifier::builder()` and `WebPkiClientVerifier::builder()` must now ensure that the crate's features are unambiguous or explicitly select a process-level provider using `CryptoProvider::install_default()`. Otherwise, these calls will panic with:
  > `no process-level CryptoProvider available -- call CryptoProvider::install_default() before this point`
  
  **We recommend that libraries rely on the process-level provider by default**, and that applications use this new API to select the provider they wish to use.

- **New unbuffered API.** [`UnbufferedClientConnection`] and [`UnbufferedServerConnection`] offer a [low-level, event-driven API which does not internally buffer data][unbuffered].
  Thanks to the team from Ferrous Systems.

- **New `no_std` support.**  A new (enabled by default) `std` crate feature now gates all APIs that depend on `std`. The above [unbuffered] APIs must be used for `no_std` support. Note that `alloc` continues to be required. Work is ongoing to reintroduce certain APIs for `no_std` users (see #1688) -- please file issues for other `no_std` use cases.
  Thanks to the team from Ferrous Systems.

- **Performance improvement: internal copying while sending data is reduced.**
  Thanks to the team from the Sōzu project.

- **Performance improvement: `write_vectored` now produces less on-the-wire overhead**, which will dramatically improve throughput if it is used with a large number of small messages.
  Thanks to the team from the Sōzu project.
  
- **`Acceptor` API error handling improvement**. If a TLS alert should be sent to inform the peer of a connection failure, this is now made available in the `Err()` variant returned from [`Acceptor::accept`] and [`Accepted::into_connection`] (which is also a **breaking change**).  Applications should write this data to the peer. See the [server_acceptor] example.

- **Support for FFDHE key exchange**: custom `CryptoProviders` can now support FFDHE key exchange, in accordance with [RFC7919]. Note that the default providers do not do this.
  Thanks to the team from Fortanix.

- **Support for servers requiring `extended_master_secret` support** from clients. See [`ServerConfig::require_ems`].
  Thanks to the team from Fortanix.

- **Extension ordering in ClientHello messages are now randomised** as an anti-fingerprinting measure. We do not foresee any interoperability issues [as Chrome has already rolled out the same change][chrome-ext-order].
  Thanks to @GomesGoncalo.
  
- **Breaking change**: `CipherSuiteCommon::integrity_limit` field removed (this was QUIC-specific, it has moved to `quic::PacketKey::integrity_limit()`).

- **Breaking change**: `crypto::cipher::BorrowedPlainMessage` and `crypto::cipher::OpaqueMessage` have been renamed (to `OutboundPlainMessage` and `OutboundOpaqueMessage`) and altered to support performance improvements.  See [the example code][provider-example-aead].

- **Breaking change**: all protocol enum types (eg. [`CipherSuite`]) have had their `get_u8`/`get_u16` accessor removed; use `u8::from()` / `u16::from()` instead.

[`aws-lc-rs`]: https://crates.io/crates/aws-lc-rs
[aws-lc-rs-reqs]: https://aws.github.io/aws-lc-rs/requirements/index.html
[fips-manual]: https://rustls.github.io/rustls/prerelease/manual/_06_fips/index.html
[aws-fips-docs]: https://github.com/aws/aws-lc-rs/tree/main/aws-lc-fips-sys#fips
[process-provider]: https://rustls.github.io/rustls/prerelease/crypto/struct.CryptoProvider.html#using-the-per-process-default-cryptoprovider
[unbuffered]: https://rustls.github.io/rustls/prerelease/unbuffered/index.html
[`UnbufferedClientConnection`]: https://rustls.github.io/rustls/prerelease/client/struct.UnbufferedClientConnection.html
[`UnbufferedServerConnection`]: https://rustls.github.io/rustls/prerelease/server/struct.UnbufferedServerConnection.html
[provider-example-aead]: https://github.com/rustls/rustls/blob/main/provider-example/src/aead.rs
[server_acceptor]: https://github.com/rustls/rustls/blob/main/examples/src/bin/server_acceptor.rs
[`CipherSuite`]: https://rustls.github.io/rustls/prerelease/enum.CipherSuite.html
[`Acceptor::accept`]: https://rustls.github.io/rustls/prerelease/server/struct.Acceptor.html#method.accept
[`Accepted::into_connection`]: https://rustls.github.io/rustls/prerelease/server/struct.Accepted.html#method.into_connection
[RFC7919]: https://datatracker.ietf.org/doc/html/rfc7919
[`ServerConfig::require_ems`]: https://rustls.github.io/rustls/prerelease/server/struct.ServerConfig.html#structfield.require_ems
[chrome-ext-order]: https://chromestatus.com/feature/5124606246518784

ctz commented:
Note that I tried using `cargo semver-checks --release-type minor` to get a list of breaking changes. It got 3 of these, but missed the last one. I am sure I've missed others.

cpu commented:
Thanks! Here's some initial feedback (sorry, can't do a "real" review for PR description content). 

> Support for process-wide selection of CryptoProviders. See [the documentation](https://rustls.github.io/rustls/prerelease/crypto/struct.CryptoProvider.html#using-the-per-process-default-cryptoprovider).

I think we should explicitly call out that code that was using `ClientConfig::builder()` or `ServerConfig::builder()` before will now see a "no process-level `CryptoProvider` available. call `CryptoProvider::install_default()` before this point" error unless they're switched to explicitly choose a provider or set the process default. I think this will be very common and since the error happens at runtime and not compile time will bite folks later than other breaking changes might.

> The above [unbuffered](https://rustls.github.io/rustls/prerelease/unbuffered/index.html) APIs must be used for no_std support.

Worth mentioning this disables some functionality that will be restored in follow-up work?

> Performance improvement: internal copying during sending data is reduced.

Nit: I'd say "while sending data" instead of "during sending data"

> "If a TLS should be sent to inform the peer of a connection failure"

Typo: "If a TLS **alert** ..."

>  Applications should write this data to the peer.

We've received feedback in the past that it's helpful to point to examples. Maybe put a forward pointer here to the server acceptor example?

> Acceptor API error handling improvement.
> ...
> Breaking change: the Result::Err() contents for Acceptor::accept and Accepted::into_connection 

I think these are the same change? Could probably fold them both into one listing.

cpu commented:
> Default cryptography provider changed to aws-lc-rs. Note that this has some implications on [platform support and build-time tool requirements][aws-lc-rs-reqs)]

Maybe worth mentioning a couple of the new reqs up front? I'm thinking the Windows nasm one in particular since I've had to update several Rustls CI configurations to add that on Windows.

ctz commented:
OK, I think I have addressed those comments. Thanks!

cpu commented:
Here are a few more items for consideration:

* https://github.com/rustls/rustls/pull/1784 It's not a breaking change per-say, but does seem like a notable bug-fix.
* https://github.com/rustls/rustls/pull/1702 seems worth a mention and a shout-out to the Fortanix folks that authored it. I don't _think_ it had any breaking changes, but might be worth a double check.
* https://github.com/rustls/rustls/pull/1754 is another new feature from the Fortanix folks not mentioned presently.
* https://github.com/rustls/rustls/pull/1730 - randomizing client hello extension ordering - since there was some concern about interoperability it seems worth an explicit mention.
* https://github.com/rustls/rustls/pull/1706 - support for P512-SHA512 signatures w/ aws-lc-rs
* https://github.com/rustls/rustls/pull/1708 - `RootCertStore` has a `FromIterator` impl now (maybe not that notable?)
* https://github.com/rustls/rustls/pull/1674 - we backported this to 0.21, but I think it's worth mentioning in this release too (?)
* https://github.com/rustls/rustls/pull/1668 - the fix for removing TLS 1.2 session tickets on error, also backported to 0.21.

djc commented:
For stuff that landed in 0.21 releases, it doesn't seem very useful to call them out (assuming they were called out appropriately in the earlier releases) -- I think people would intuitively expect them to be part of this release anyway. I also agree with your parenthetical that the `FromIterator` addition probably isn't worth calling out, but I agree the other ones should probably be included.

djc commented:
> # Release notes (proposed)
> 
>     * **Default cryptography provider changed** to [`aws-lc-rs`](https://crates.io/crates/aws-lc-rs). Note that this has some implications on [platform support and build-time tool requirements](https://aws.github.io/aws-lc-rs/requirements/index.html) such as `cmake` on all platforms and `nasm` on Windows.
>       Support for `ring` continues to be available: set the `ring` crate feature.
> 
>     * **Support for FIPS validated mode** with [`aws-lc-rs`](https://crates.io/crates/aws-lc-rs): see [the manual section](https://rustls.github.io/rustls/prerelease/manual/_06_fips/index.html) and [aws-lc-rs's FIPS documentation](https://github.com/aws/aws-lc-rs/tree/main/aws-lc-fips-sys#fips). Note that `aws-lc-rs` in FIPS mode has further build-time requirements.

This evokes the question for a link to details on the "further build-time requirements", maybe just mention explicitly "as detailed in the FIPS documentation."?

>     * **Support for process-wide selection of `CryptoProvider`s**.  See [the documentation](https://rustls.github.io/rustls/prerelease/crypto/struct.CryptoProvider.html#using-the-per-process-default-cryptoprovider). Note that callers of `ClientConfig::builder()`, `ServerConfig::builder()`, `WebPkiServerVerifier::builder()` and `WebPkiClientVerifier::builder()`, must now ensure that the crate's features are unambigious or explicitly select a process-level provider using `CryptoProvider::install_default()`. Otherwise, these calls will panic with "no process-level `CryptoProvider` available. call `CryptoProvider::install_default()` before this point".

I would probably omit the comma before "must now" but this maybe inviting some age-old discussion on the Oxford comma, not sure? Also "unambigious" -> "unambiguous" and not sure what is going on with the last sentence here, which isn't capitalized and ends with a stray `"`? I also think we should include some guidance here suggesting that libraries depending on rustls 0.23+ should probably rely on the process default, allowing the application to install one.

>     * **New unbuffered API.** [`UnbufferedClientConnection`] and [`UnbufferedServerConnection`] offer a [low-level, event-driven API which does not internally buffer data](https://rustls.github.io/rustls/prerelease/unbuffered/index.html).
>       Thanks to the team from Ferrous Systems.
> 
>     * **New `no_std` support.**  A new `std` crate feature now gates all APIs that depend on `std`, this is a default feature. The above [unbuffered](https://rustls.github.io/rustls/prerelease/unbuffered/index.html) APIs must be used for `no_std` support. Note that `alloc` continues to be required. There is some future work to reintroduce some APIs gated under the `std` feature.
>       Thanks to the team from Ferrous Systems.

Would replace ", this is a default feature" with "A new (enabled by default) `std` crate feature [..]". "There is some future work [..]" feels a bit informal/messy, maybe something like "In order to merge `no_std` support, some features that could be supported without `std` support are currently compiled when `std` support is disabled. Work is ongoing to re-introduce support for some of these features -- please file issues for `no_std` use cases that need additional work."?

>     * **Performance improvement: `write_vectored` now produces less on-the-wire overhead**, which will dramatically improve throughput if it is used with a large number of small messages.
>       Thanks to the team from the Sōzu project.
> 
>     * **Performance improvement: internal copying while sending data is reduced.**
>       Thanks to the team from the Sōzu project.

Given the wider applicability, should this go before the other one?

>     * **`Acceptor` API error handling improvement**. If a TLS alert should be sent to inform the peer of a connection failure, this is now made available in the `Err()` variant returned from [`Acceptor::accept`](https://rustls.github.io/rustls/prerelease/server/struct.Acceptor.html#method.accept) and [`Accepted::into_connection`](https://rustls.github.io/rustls/prerelease/server/struct.Accepted.html#method.into_connection) (which is also a **breaking change**).  Applications should write this data to the peer. See the [server_acceptor](https://github.com/rustls/rustls/blob/main/examples/src/bin/server_acceptor.rs) example.
> 
>     * **Breaking change**: `CipherSuiteCommon::integrity_limit` field removed (this was QUIC-specific, it has moved to `quic::PacketKey::integrity_limit()`).
> 
>     * **Breaking change**: `crypto::cipher::BorrowedPlainMessage` and `crypto::cipher::OpaqueMessage` have been renamed (to `OutboundPlainMessage` and `OutboundOpaqueMessage`) and altered to support performance improvements.  See [the example code](https://github.com/rustls/rustls/blob/main/provider-example/src/aead.rs).
> 
>     * **Breaking change**: all protocol enum types (eg. [`CipherSuite`](https://rustls.github.io/rustls/prerelease/enum.CipherSuite.html)) have had their `get_u8`/`get_u16` accessor removed; use `u8::from()` / `u16::from()` instead.

djc commented:
Should we also call out the public availability of rustls-platform-verifier somehow?

ctz commented:
> This evokes the question for a link to details on the "further build-time requirements", maybe just mention explicitly "as detailed in the FIPS documentation."?

Done.

> I would probably omit the comma before "must now" but this maybe inviting some age-old discussion on the Oxford comma, not sure? Also "unambigious" -> "unambiguous" and not sure what is going on with the last sentence here, which isn't capitalized and ends with a stray `"`? I also think we should include some guidance here suggesting that libraries depending on rustls 0.23+ should probably rely on the process default, allowing the application to install one.

Addressed these points.

The last sentence was a quote of the panic message: have blockquoted this to take it out of the flow of the text.

> Would replace ", this is a default feature" with "A new (enabled by default) `std` crate feature [..]". "There is some future work [..]" feels a bit informal/messy, maybe something like "In order to merge `no_std` support, some features that could be supported without `std` support are currently compiled when `std` support is disabled. Work is ongoing to re-introduce support for some of these features -- please file issues for `no_std` use cases that need additional work."?

Have dropped the "future work" sentence & adapted the "work is ongoing" part.

> Given the wider applicability, should this go before the other one?

Swapped.

djc commented:
> The last sentence was a quote of the panic message: have blockquoted this to take it out of the flow of the text.

I guess I would suggest we use different (from `.`) punctuation inside the panic message, maybe `--` or `;`?

ctz commented:
> Here are a few more items for consideration:
> 
> * [Correct kx group selection #1784](https://github.com/rustls/rustls/pull/1784) It's not a breaking change per-say, but does seem like a notable bug-fix.

I think this is functionally invisible really.

> * [Add support for FFDHE key exchange #1702](https://github.com/rustls/rustls/pull/1702) seems worth a mention and a shout-out to the Fortanix folks that authored it. I don't _think_ it had any breaking changes, but might be worth a double check.
> * [Add config option for requiring peer must support `extended_master_secret` extension in TLS 1.2 #1754](https://github.com/rustls/rustls/pull/1754) is another new feature from the Fortanix folks not mentioned presently.
> * [Randomize ClientHello extensions #1730](https://github.com/rustls/rustls/pull/1730) - randomizing client hello extension ordering - since there was some concern about interoperability it seems worth an explicit mention.

Added these.

> * [Support P521-SHA512 signatures with aws-lc-rs #1706](https://github.com/rustls/rustls/pull/1706) - support for P512-SHA512 signatures w/ aws-lc-rs
> * [Implement FromIterator for RootCertStore #1708](https://github.com/rustls/rustls/pull/1708) - `RootCertStore` has a `FromIterator` impl now (maybe not that notable?)
> * [Re-export the types crate contents #1674](https://github.com/rustls/rustls/pull/1674) - we backported this to 0.21, but I think it's worth mentioning in this release too (?)
> * [remove the TLS 1.2 session ticket on DecryptError #1668](https://github.com/rustls/rustls/pull/1668) - the fix for removing TLS 1.2 session tickets on error, also backported to 0.21.

I think these are in 0.22.1 or 0.22.2.

cpu commented:
I did another read through of the proposed text and it looks good to me :+1:
